From a6fce70f6c4e4d355c14475be8f5669a83c7023f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Luis=20Javier=20Merino=20Mor=C3=A1n?= <ninjalj@gmail.com>
Date: Sat, 30 Apr 2022 19:12:36 +0200
Subject: [PATCH] Fix scroll position jumps regression

Commit d37d3ac1 "CompactHistoryScroll: Remove _maxLineCount + 5 lines at
a time" caused a regression: addHistLine queried the number of lines in
history before and after adding a line, and if it had not incremented it
assumed one line entered history and another was dropped at the other
end.  Now, lines are dropped from history in batches, so take care of
that.

Very similar to the regression fixed at 7a1e4768, only this one is about
keeping the scroll position on the presence of scroll, and that other
one is about keeping the selection.

Thanks to the bug reporters Michael and Luke-Jr for the heads-up.

BUG: 452955
BUG: 453112
(cherry picked from commit d2ca202abd4e6b415a7647927db493337092d5f3)
---
 src/Screen.cpp | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/Screen.cpp b/src/Screen.cpp
index 66b47535c..e57314cd5 100644
--- a/src/Screen.cpp
+++ b/src/Screen.cpp
@@ -1735,11 +1735,11 @@ void Screen::addHistLine()
 
         // If the history is full, increment the count
         // of dropped _lines
-        if (newHistLines == oldHistLines) {
-            ++_droppedLines;
+        if (newHistLines <= oldHistLines) {
+            _droppedLines += oldHistLines - newHistLines + 1;
 
-            // We removed a line, we need to verify if we need to remove a URL.
-            _escapeSequenceUrlExtractor->historyLinesRemoved(1);
+            // We removed some lines, we need to verify if we need to remove a URL.
+            _escapeSequenceUrlExtractor->historyLinesRemoved(oldHistLines - newHistLines + 1);
         }
     }
 
-- 
GitLab

